Fix #5911, Fix #5907, Fix #5904, Fix #5903, Fix #5902, Fix #5901, Fix #5914: Onboarding M1 Fixes

adhiamboperes opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fix #5911
The new app language screen should share functionality & UI components and styling with the new learner onboarding flow, featuring a dopdown instead of the current radio buttons.

The UI has been refactored to share the same UI as the onboarding screen's app language selection. The onboarding text and navigation buttons have been removed. Going back from the language selection screens to the options screen require the use of the system or toolbar back buttons, maintaining consistency with the legacy experience.

https://github.com/user-attachments/assets/0c0ec2bc-f2c6-4f9c-954a-307943d00c54


Fix #5907 
In [AppStartupStateController.kt](https://github.com/oppia/oppia-android/blob/273207fbbd06c10c824d819501b3f7009d233621/domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt#L65-L74), https://github.com/oppia/oppia-android/blob/273207fbbd06c10c824d819501b3f7009d233621/domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt#L65-L74, markOnboardingFlowCompleted does not notify observers that onboarding has been completed.

This caused a bug where exiting the app after selecting parent/teacher flow via back button and reopening the app brings the user back to the beginning of the onboarding flow even if they logged into the admin account.
This is tested manually.

https://github.com/user-attachments/assets/b6f5def9-8ebd-42ff-ae5b-9ce791dbea1a

Fix #5904
Changes have been made so that the Step 5 of 5 text  should not be visible if the audio selection screen is opened from options.

https://github.com/user-attachments/assets/b1267a52-ffff-4cbb-acae-6b79d2d66293

Fix #5901, Fix #5902, Fix #5903 
The audio language did not set correctly in several instances because the user was not logged in beforehand. Logging in the user fixes this. Additional fixes were made such that the fallbacks and pre-selections are handled correctly.

1. Opening the options screen always shows 'English' rather than the setting actually used by the learner.

https://github.com/user-attachments/assets/de3342b2-a203-4f2c-9d81-80e117253482

2. Attempting to change the user's audio language via options overwrites the initial default setting back to English after selecting anything from the Options menu (but works once the profile is re-entered )

https://github.com/user-attachments/assets/571a093e-97d1-4889-bf24-79b9cb14cd2e

3. When setting the audio language for a migrated non-admin profile, it simply doesn't set at all.

Demo with feature flag off:

https://github.com/user-attachments/assets/450eddc2-1d19-4870-858d-dba0211ec1c0

Demo with feature flag on:

https://github.com/user-attachments/assets/cb58d27c-bb2a-4204-9ba1-04caa1ecb297

4. Attempting to change the user's audio language via options fails immediately after onboarding.

https://github.com/user-attachments/assets/95865fc2-824d-4516-9246-929cc99f0a3e

Fix #5914: Where the user could bypass the profile login screen in a migrated profile due to the onboarding flow kicking in.

Creating the profile:

https://github.com/user-attachments/assets/dcb21330-827a-4fdc-b3f6-16e75119eb93

Demo of fix:

https://github.com/user-attachments/assets/c0f1ac27-2808-49f4-9bed-e9c9a46e79be

## Other changes
**Removed duplicate tests**
`testProfileOnboardingState_existingProfilesV1_returnsUnknownProfileTypeMode` is a duplicate of `testProfileOnboardingState_oneAdminProfileWithPin_returnsAdminOnlyMode` not that fetching the profiles list updates profiles

**Refactored the use of Observer**
In [PinPasswordActivityPresenter.kt](https://github.com/oppia/oppia-android/pull/5908/files/9e0413c41667827218fcb889c12119e49ea8aba9#diff-3e555282c55b1efddd1b42689f0924012a9ae8e795c36fa8e0b96e4c57f239ab), Kotlin threw the warning:

```kotlin
warning: candidate resolution will be changed soon, please use fully qualified name to invoke the following closer candidate explicitly '
public open fun observe(p0: LifecycleOwner, p1: Observer<in AsyncResult<Any?>!>): Unit defined in androidx.lifecycle.LiveData'
              .loginToProfile(profileId).toLiveData().observe(
```

Since -WError is enabled, the project could not build. I refactored occurances of

`loginToProfile(profileId).toLiveData().observe()`

to

```
loginToProfile(profileId).toLiveData().observe(lifecycleOwner, Observer { result ->
  // handle result
})
```

in order to fix the warning.

**Add Profile Button Bug**
I discovered a bug where, once the `add profile` button has been hidden when max profiles have been created, it did not show up again when some profiles are deleted, until the user exits and relaunches the app. I applied a fix and a test.

**Downgrade experience**
It was reported that the onboarding experience shows again if the app launched with the feature flag on, and at least one user completed onboarding, and thereafter was relaunched with the feature flag off(legacy version). I could not repro this. Below is demo proof that the onboarding flow does not show up on downgrade:

Feature Flag On:

https://github.com/user-attachments/assets/bb575e45-8355-4f3d-b9c3-54f47e10dc79

Feature Flag Off:

https://github.com/user-attachments/assets/7e8de2fa-f303-4669-b1a1-d30aeaf831a2



## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-A11y-Guide))
- For PRs introducing new UI elements or color changes, both light and dark mode screenshots must be included
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing


> I have included the video demos inline for clarity.

adhiamboperes committed:
Ensure sole learner home screen language is correct.

adhiamboperes committed:
Fix audio language issues

adhiamboperes committed:
Fix migration issues

adhiamboperes committed:
Fix security issue

adhiamboperes committed:
Fix navigation issues

adhiamboperes committed:
Rename presenter

adhiamboperes committed:
Support new app language selection UI

adhiamboperes committed:
Add audio language selection tests

adhiamboperes committed:
Add audio language selection tests

adhiamboperes committed:
Clean up OnboardingFragmentTest

adhiamboperes commented on app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt:
```diff
@@ -1,69 +1,170 @@
 import javax.inject.Inject
 
-/** The presenter for [AppLanguageFragment]. */
 class AppLanguageFragmentPresenter @Inject constructor(
```
Add a kdoc

adhiamboperes commented on app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt:
```diff
@@ -1,69 +1,170 @@
 
-  /** Initializes and creates the views for [AppLanguageFragment]. */
-  fun handleOnCreateView(
+  /** Handle creation and binding of the [AppLanguageFragment] layout. */
```
```suggestion
  /** Handles the creation and binding of the [AppLanguageFragment] layout. */
```

adhiamboperes replied:
Done

adhiamboperes commented on app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt:
```diff
@@ -1,69 +1,170 @@
     inflater: LayoutInflater,
     container: ViewGroup?,
+    outState: Bundle?,
     prefSummaryValue: OppiaLanguage,
```
```suggestion
    currentSelection: OppiaLanguage,
```

adhiamboperes replied:
renamed

adhiamboperes commented on app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt:
```diff
@@ -1,69 +1,170 @@
+        )
+      }
+
+      onboardingLanguageDropdown.apply {
```
```suggestion
      appLanguageDropdown.apply {
```

adhiamboperes commented on app/src/main/java/org/oppia/android/app/options/AppLanguageFragmentPresenter.kt:
```diff
@@ -1,69 +1,170 @@
+          R.id.onboarding_language_text_view,
+          languagesList.map { appLanguageResourceHandler.computeLocalizedDisplayName(it) }
+        )
+        onboardingLanguageDropdown.setAdapter(adapter)
```
```suggestion
          R.layout.appg_language_dropdown_item,
          R.id.app_language_text_view,
          languagesList.map { appLanguageResourceHandler.computeLocalizedDisplayName(it) }
        )
        appLanguageDropdown.setAdapter(adapter)
```

adhiamboperes replied:
Refactored.

adhiamboperes committed:
Add tests for app language selection

adhiamboperes committed:
Merge branch 'onboarding-m1-fixes' of github.com:oppia/oppia-android into onboarding-m1-fixes

adhiamboperes committed:
Fix static check failures

adhiamboperes committed:
Fix options tests

adhiamboperes commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -226,11 +235,14 @@ class PinPasswordActivityPresenter @Inject constructor(
-          confirmedDeletion = true
-          activity.finishAffinity()
-        }
+        profileManagementController.deleteAllProfiles().toLiveData().observe(
+          activity,
+          Observer {
+            // Regardless of the result of the operation, always restart the app.
```
Document this change

adhiamboperes replied:
Done

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/options/AppLanguageFragmentTest.kt:
```diff
@@ -254,13 +279,241 @@ class AppLanguageFragmentTest {
+//        matches(withText(R.string.arabic_localized_language_name))
+//      )
+//    }
+//  }
```
Remove commented out code

adhiamboperes replied:
Removed

adhiamboperes commented on app/src/main/java/org/oppia/android/app/profile/ProfileChooserFragmentPresenter.kt:
```diff
@@ -286,11 +286,27 @@ class ProfileChooserFragmentPresenter @Inject constructor(
+  private fun computeLoginRoute(profile: Profile) {
+    when {
+      profile.profileType == ProfileType.SUPERVISOR -> { routeToHomeOrPinScreens(profile) }
+      !profile.completedProfileOnboarding -> { routeToOnboardingOrPinScreens(profile) }
```
```suggestion
      profile.startedProfileOnboarding -> { routeToOnboardingOrPinScreens(profile) }
```

Is likely more explicit

adhiamboperes replied:
This actually means a different thing in this scenario. Migrated profiles will nevers have profile.startedProfileOnboarding == true, so keeping the original.

adhiamboperes committed:
Fix navigation tests

adhiamboperes reviewed (commented):
Self reviewed

adhiamboperes committed:
Fix profile chooser failures

adhiamboperes committed:
Fix app startup failures

adhiamboperes committed:
Fix app profile management controller failures

adhiamboperes committed:
Fix self review comments

adhiamboperes committed:
Fix typo

adhiamboperes committed:
Fix test

adhiamboperes committed:
revert change

adhiamboperes committed:
revert changes

adhiamboperes committed:
revert test changes

adhiamboperes committed:
fix bugs and UI flows

BenHenning commented on app/src/main/java/org/oppia/android/app/options/AudioLanguageSelectionViewModel.kt:
```diff
@@ -89,16 +90,27 @@ class AudioLanguageSelectionViewModel @Inject constructor(
+    ) { profileAudioLanguage, appLanguageSelection ->
+      if (profileAudioLanguage != OppiaLanguage.LANGUAGE_UNSPECIFIED ||
+        profileAudioLanguage != OppiaLanguage.UNRECOGNIZED
+      )
```
Nit: Add braces since the whole `if` doesn't fit on one line.

adhiamboperes replied:
Done.

adhiamboperes replied:
This code has moved to the `TranslationController`.

BenHenning commented on app/src/main/java/org/oppia/android/app/options/AudioLanguageSelectionViewModel.kt:
```diff
@@ -89,16 +90,27 @@ class AudioLanguageSelectionViewModel @Inject constructor(
+      appLanguageSelectionProvider,
+      PROFILE_LANGUAGE_PROVIDER_ID
+    ) { profileAudioLanguage, appLanguageSelection ->
+      if (profileAudioLanguage != OppiaLanguage.LANGUAGE_UNSPECIFIED ||
```
This logic probably should be in the domain layer since it's the general fallback logic for languages and should be ensured to be consistent across the entire app rather than just for this UI.

Also per the TODO above, is this actually addressing #4938?

adhiamboperes replied:
I have moved the logic to the TranslationController and added new tests.

The TODO was added because part of #4938's requirements is to have new learner profiles default to Admin's audio language(next milestone).

BenHenning commented on domain/src/main/java/org/oppia/android/domain/onboarding/AppStartupStateController.kt:
```diff
@@ -118,7 +113,7 @@ class AppStartupStateController @Inject constructor(
     // in-memory (which means it will be inadvertently overwritten when updating onboarding state
     // here).
-    val deferred = onboardingFlowStore.storeDataAsync(updateInMemoryCache = false) { state ->
+    val deferred = onboardingFlowStore.storeDataAsync { state ->
```
Is this still true? I'm wondering if the implementation can be simplified as this comment removed.

adhiamboperes replied:
It isn't. I refactored and tested.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/ProfileChooserViewModel.kt:
```diff
@@ -84,6 +84,8 @@ class ProfileChooserViewModel @Inject constructor(
 
     if (sortedProfileList.size == 10) {
       canAddProfile.set(false)
+    } else {
+      canAddProfile.set(true)
     }
```
```suggestion
    canAddProfile.set(sortedProfileList.size < 10)
```

Can be simplified a bit.

adhiamboperes replied:
Done, thanks!

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -257,4 +269,38 @@ class PinPasswordActivityPresenter @Inject constructor(
+  }
+
+  private fun launchOnboardingScreen(profileId: ProfileId, profileName: String) {
+    val introActivityParams = IntroActivityParams.newBuilder()
+      .setProfileNickname(profileName)
+      .setParentScreen(IntroActivityParams.ParentScreen.PIN_PASSWORD_SCREEN)
+      .build()
+
+    val intent = IntroActivity.createIntroActivity(activity)
```
This seems a bit odd to me--shouldn't `createIntroActivity` be properly setting up its intent?

adhiamboperes replied:
Yes - I have done a refactor so that we are passing the profileId and arguments to the Intent helper instead.

BenHenning commented on app/src/main/java/org/oppia/android/app/profile/PinPasswordActivityPresenter.kt:
```diff
@@ -257,4 +269,38 @@ class PinPasswordActivityPresenter @Inject constructor(
+
+  private fun launchHomeScreen() {
+    activity.startActivity(
+      if (enableMultipleClassrooms.value)
+        ClassroomListActivity.createClassroomListActivity(activity, profileId)
+      else
+        HomeActivity.createHomeActivity(activity, profileId)
```
Nit: needs braces since it's not fitting on a single line (though the else probably could be reformatted that way).

adhiamboperes replied:
Added braces for both branches.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt:
```diff
@@ -246,89 +248,178 @@ class AudioLanguageFragmentTest {
-        matches(withEffectiveVisibility(Visibility.VISIBLE))
-      )
-    }
+    launchV2FlowWithLanguage(ENGLISH_AUDIO_LANGUAGE, ParentScreen.LEARNER_INTRO_SCREEN)
```
Can you perhaps import `LEARNER_INTRO_SCREEN` so that this & the other tests don't require the extra indentation? It's a bit hard to review this test with the extra, possibly avoidable whitespace changes.

adhiamboperes replied:
Done. 

I think I hit reformat last minute and caused this indent issue

BenHenning reviewed (changes requested):
Thanks @adhiamboperes! Took a full pass except for `AudioLanguageFragmentTest` since I'm hoping some reformatting will make that a bit easier to review.

One question: is the new audio language screen automatically closing when accessed via options? That doesn't seem like a consistent user experience with language selection. I'd expect the drop-down to update the language but otherwise require the user to use the back arrow or back navigation to navigate back to the options screen.

In fact, shouldn't that also be true for the regular onboarding flow? That is, they should have to click a button to continue rather than the selection automatically progressing navigation. It's possible I misunderstood the videos, though.

adhiamboperes committed:
address reviewer comment

adhiamboperes commented on domain/src/main/java/org/oppia/android/domain/translation/TranslationController.kt:
```diff
@@ -379,6 +411,7 @@ class TranslationController @Inject constructor(
     return when (languageSelection.selectionTypeCase) {
       AppLanguageSelection.SelectionTypeCase.SELECTED_LANGUAGE ->
         LanguageResolutionStatus.Resolved(languageSelection.selectedLanguage)
+
```
Remove newlines from file

adhiamboperes commented on app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt:
```diff
@@ -439,29 +499,45 @@ class AudioLanguageFragmentTest {
         )
 
         onView(withId(R.id.onboarding_navigation_continue)).perform(click())
+
```
Remove newline

adhiamboperes committed:
minor cleanup

adhiamboperes reviewed (commented):
Self reviewed latest changes. LGTM.

adhiamboperes commented:
Thanks @BenHenning! i have addressed all your comments.
 
> One question: is the new audio language screen automatically closing when accessed via options? That doesn't seem like a consistent user experience with language selection. I'd expect the drop-down to update the language but otherwise require the user to use the back arrow or back navigation to navigate back to the options screen.

No, the screen remains open, so that they can make different selections via the dropdown(both audio and app language slections).

> In fact, shouldn't that also be true for the regular onboarding flow? That is, they should have to click a button to continue rather than the selection automatically progressing navigation. It's possible I misunderstood the videos, though.

Yes, during onboarding, there is a continue button. but in options, the back navigation is via the system or back navigation buttons.

I have re-recorded the audio language selectin demo videos. PTAL.

BenHenning commented on domain/src/test/java/org/oppia/android/domain/translation/TranslationControllerTest.kt:
```diff
@@ -1371,6 +1371,183 @@ class TranslationControllerTest {
+  }
+
+  @Test
+  fun testGetAudioLanguagePreselection_IndiaDefaultLocale_returnsHindiLanguageAsPreselection() {
```
```suggestion
  fun testGetAudioLanguagePreselection_indiaDefaultLocale_returnsHindiLanguageAsPreselection() {
```

Nit on snake case capitalization.

adhiamboperes replied:
Fixed.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt:
```diff
@@ -290,12 +290,70 @@ class AudioLanguageFragmentTest {
+  @Test
+  fun testFragment_landscapeMode_backButtonPressed_currentScreenIsDestroyed() {
+    launchV2FlowWithLanguage(ENGLISH_AUDIO_LANGUAGE, LEARNER_INTRO_SCREEN).use { scenario ->
+      rotateToLandscape()
```
Nit here & elsewhere in this file: it's probably better to start the test in landscape mode for landscape (e.g. using test `@Config` qualifiers) tests since rotation doesn't actually change the orientation correctly currently (#2277 tracks this, though perhaps not well). The orientation change flow is still useful for verifying saved instance state cases, though.

adhiamboperes replied:
Changed the relevant instances.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/options/AudioLanguageFragmentTest.kt:
```diff
@@ -455,13 +514,29 @@ class AudioLanguageFragmentTest {
+
+        testCoroutineDispatchers.runCurrent()
+
+        onView(withId(R.id.audio_language_dropdown_list)).check(
+          matches(withText(R.string.nigerian_pidgin_localized_language_name))
+        )
```
Does this actually validate the default and not any of the drop-down items? It's a bit hard for me to check just by reading the assertion, so thought I'd double check.

adhiamboperes replied:
No, this is meant to validate that updating the language works. I have added a test that verifies defaults.

BenHenning reviewed (approved):
Thanks @adhiamboperes! The PR largely LGTM. I have a few follow-up comments, but I'm happy for you to resolve them if you don't want to wait for another review. Otherwise, I can take a look.

adhiamboperes committed:
minor changes

adhiamboperes commented:
Please take a look @BenHenning.

BenHenning reviewed (approved):
Thanks @adhiamboperes! LGTM.

adhiamboperes committed:
Merge branch 'develop' of github.com:oppia/oppia-android into onboarding-m1-fixes
